bundles/org.openhab.binding.saicismart/src/main/java/org/openhab/binding/saicismart/internal/SAICiSMARTHandler.java

```diff
@@ -0,0 +1,278 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.i18n.TimeZoneProvider;
+import org.openhab.core.library.types.DateTimeType;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.thing.binding.builder.ThingBuilder;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.GsonBuilder;
+
+import net.heberling.ismart.abrp.ABRP;
+import net.heberling.ismart.asn1.v1_1.entity.Message;
+import net.heberling.ismart.asn1.v2_1.MessageCoder;
+import net.heberling.ismart.asn1.v2_1.entity.OTA_RVCReq;
+import net.heberling.ismart.asn1.v2_1.entity.OTA_RVCStatus25857;
+import net.heberling.ismart.asn1.v2_1.entity.OTA_RVMVehicleStatusResp25857;
+import net.heberling.ismart.asn1.v2_1.entity.RvcReqParam;
+import net.heberling.ismart.asn1.v3_0.entity.OTA_ChrgMangDataResp;
+
+/**
+ * The {@link SAICiSMARTHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Markus Heberling - Initial contribution
+ */
+@NonNullByDefault
+public class SAICiSMARTHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(SAICiSMARTHandler.class);
+
+    private final TimeZoneProvider timeZoneProvider;
+
+    @Nullable
+    SAICiSMARTVehicleConfiguration config;
+    private @Nullable Future<?> pollingJob;
+    private ZonedDateTime lastAlarmMessage;
+    private ZonedDateTime lastCarActivity;
+
+    /**
+     * If the binding is initialized, treat the car as active (lastCarActivity = now) to get some first data.
+     * 
+     * @param httpClientFactory
+     * @param timeZoneProvider
+     * @param thing
+     */
+    public SAICiSMARTHandler(TimeZoneProvider timeZoneProvider, Thing thing) {
+        super(thing);
+        this.timeZoneProvider = timeZoneProvider;
+        lastAlarmMessage = ZonedDateTime.now(getTimeZone());
+        lastCarActivity = ZonedDateTime.now(getTimeZone());
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (channelUID.getId().equals(SAICiSMARTBindingConstants.CHANNEL_FORCE_REFRESH) && command == OnOffType.ON) {
+            // reset channel to off
+            updateState(CHANNEL_FORCE_REFRESH, OnOffType.from(false));
+            // update internal activity date, to query the car for about a minute
+            notifyCarActivity(ZonedDateTime.now(getTimeZone()).minus(9, ChronoUnit.MINUTES), true);
```

Comment:
How does this work as it looks like it has 9 minutes hardcoded. If this is related to the 10 minutes on line 152, i suggest to make that one on line 152 a class level constant and add it here like  `UPDATE_INTERVAL_MINUTES - 1` (Not sure if that makes sense.)

Reply:
From the forum there is an explanation of the polling.

"There is a channel “Last Car Activity”. As long as the current time is no later than 15 Minutes after the time in that channel, the binding does updates basically as fast as it can.

The car is marked as active, if either the engine runs, the car is charging or remote climate is activated. This are basically the states of the car when the H/V battery is on.

This works as long as the car is active, but as soon as it not active anymore I don’t do any polling anymore. You can enable start notifications in the iSmart App, that send a message over the API every time the engine is started. I’m polling those and set the car to active to start car polling again.

The “Force Refresh” button sets the active date to now - 14 Minutes, so that the fast polling is enabled for a minute. This can be used to wake up polling, if for example a charger enables charging in the middle of the night, or you want to see the interior temperature in the morning to see if you need to defrost.

There seems to be no API limit on how fast or how often you can query the car data, but if the H/V battery is off, you can drain your 12V battery pretty fast.

I suppose it is possible to query the car once every X minutes in “inactive” mode, to get the car status for a parked car. But after getting my first 12V battery warning in the morning, I haven’t tried that again. :smiley:"

(REF: https://community.openhab.org/t/new-binding-integration-of-saic-mg-ismart-based-vehicles/140136/9?u=doug_culnane)

I tried to improve this but ended up with a low 12v battery so gave it too.  This explains the code a bit as the strategy is not initially obvious.

Reply:
SAICiSMARTBindingConstants.POLLING_ACTIVE_MINS constant to help clear this up.